marconi-tutorial/src/Marconi/Tutorial/Indexers/AddressCount.hs

```diff
@@ -0,0 +1,242 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+-- TODO Make explicit internal modules of marconi-core. Everything expect `Marconi.Core.Experiment`
+-- should be in `Marconi.Core.Experiment.Internal.*`.
+-- TODO Discenpancy between query of ListIndexer and SqliteIndexer in case where we provide
+-- ChainPointAtGenesis. Check test scenario when we provide the ChainPointAtGenesis in the
+-- generator.
+-- TODO Hard to tell what to provide as Query for singleInsertSQLiteIndexer (especially the lastSync query). Need to change `lastSyncQuery` to `lastResumablePointQuery`.
+-- TODO Change back Timed to TimedEvent
+module Marconi.Tutorial.Indexers.AddressCount where
+
+import Cardano.Api qualified as C
+import Control.Concurrent (MVar)
+import Control.Lens (at, folded, sumOf, to, (&), (.~), (^.))
+import Control.Monad.Except (MonadError)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Database.SQLite.Simple qualified as SQL
+import Database.SQLite.Simple.ToField qualified as SQL
+import GHC.Generics (Generic)
+import Marconi.ChainIndex.Orphans ()
+import Marconi.ChainIndex.Types (SecurityParam (SecurityParam))
+import Marconi.ChainIndex.Utils qualified as Utils
+import Marconi.Core.Experiment (
+  HasGenesis (genesis),
+  IndexerError,
+  ListIndexer,
+  MixedIndexer,
+  Point,
+  Queryable (query),
+  Result,
+  SQLInsertPlan (SQLInsertPlan),
+  SQLiteIndexer (SQLiteIndexer),
+  Timed (Timed),
+  event,
+  events,
+  mkListIndexer,
+  mkMixedIndexer,
+  mkSqliteIndexer,
+  point,
+ )
+import Marconi.Core.Experiment qualified as Core
+import Text.RawString.QQ (r)
+
+addressCountWorker
+  :: FilePath
+  -> SecurityParam
+  -> IO
+      ( MVar (Core.MixedIndexer Core.SQLiteIndexer Core.ListIndexer AddressCountEvent)
+      , Core.Worker (C.BlockInMode C.CardanoMode, C.EpochNo) C.ChainPoint
+      )
+addressCountWorker dbPath securityParam = do
+  let extract = getEventsFromBlock
+  ix <- Utils.toException $ mkAddressCountMixedIndexer dbPath securityParam
+  Core.createWorker (pure . extract) ix
+
+newtype AddressCountEvent = AddressCountEvent {unAddressCountEvent :: Map C.AddressAny Int}
+  deriving (Show)
+
+type instance Point AddressCountEvent = C.ChainPoint
+
+newtype AddressCountQuery = AddressCountQuery C.AddressAny
+  deriving (Eq, Ord, Show)
+
+type instance Result AddressCountQuery = Int
+
+-- Not idea that the 'Maybe' type is forced here?
```

Comment:
Let me rephrase the advantage. So if I provide `Nothing`, then this block event will _not_ be used in the batching. However, if I use `Just []`, then it's just inefficient because it adds an "empty" element in the batch?

Reply:
Yes, when we had just `event`, without maybe, we passed everything to the `inDatabase` indexer, leading to a SQL transaction even if no rows can be inserted. Also, in the general case, an event may be something else than a list.